saexec/ethtests/block_test_util.go

```diff
@@ -0,0 +1,408 @@
+		}
+	}
+	return t.validateImportedHeaders(sut.Chain, validBlocks)
+}
+
+func (t *BlockTest) genesis(config *params.ChainConfig) *core.Genesis {
+	return &core.Genesis{
+		Config:        config,
+		Nonce:         t.json.Genesis.Nonce.Uint64(),
+		Timestamp:     t.json.Genesis.Timestamp,
+		ParentHash:    t.json.Genesis.ParentHash,
+		ExtraData:     t.json.Genesis.ExtraData,
+		GasLimit:      t.json.Genesis.GasLimit,
+		GasUsed:       t.json.Genesis.GasUsed,
+		Difficulty:    t.json.Genesis.Difficulty,
+		Mixhash:       t.json.Genesis.MixHash,
+		Coinbase:      t.json.Genesis.Coinbase,
+		Alloc:         t.json.Pre,
+		BaseFee:       t.json.Genesis.BaseFeePerGas,
+		BlobGasUsed:   t.json.Genesis.BlobGasUsed,
+		ExcessBlobGas: t.json.Genesis.ExcessBlobGas,
+	}
+}
+
+/*
+See https://github.com/ethereum/tests/wiki/Blockchain-Tests-II
+
+	Whether a block is valid or not is a bit subtle, it's defined by presence of
+	blockHeader, transactions and uncleHeaders fields. If they are missing, the block is
+	invalid and we must verify that we do not accept it.
+
+	Since some tests mix valid and invalid blocks we need to check this for every block.
+
+	If a block is invalid it does not necessarily fail the test, if it's invalidness is
+	expected we are expected to ignore it and continue processing and then validate the
+	post state.
+*/
+func (t *BlockTest) insertBlocks(ctx context.Context, tb testing.TB, sut *saexectest.SUT) ([]btBlock, error) {
+	validBlocks := make([]btBlock, 0)
+	blocks := make([]*types.Block, 0)
+	// insert the test blocks, which will execute all transactions
+	for bi, b := range t.json.Blocks {
+		cb, err := b.decode()
+		if err != nil {
+			if b.BlockHeader == nil {
+				tb.Log("Block decoding failed", "index", bi, "err", err)
+				continue // OK - block is supposed to be invalid, continue with next block
+			} else {
+				return nil, fmt.Errorf("block RLP decoding failed when expected to succeed: %v", err)
+			}
+		}
+
+		if b.BlockHeader == nil {
+			// TODO(cey): We don't have any insertion rules at the moment, so skip them.
+			tb.Logf("skipping checking the invalid test block insertion (index %d) with expected exception: %v, because we don't have any insertion rules", bi, b.ExpectException)
+			tb.SkipNow()
+			if data, err := json.MarshalIndent(cb.Header(), "", "  "); err == nil {
+				fmt.Fprintf(os.Stderr, "block (index %d) insertion should have failed due to: %v:\n%v\n",
+					bi, b.ExpectException, string(data))
+			}
+			return nil, fmt.Errorf("block (index %d) insertion should have failed due to: %v",
+				bi, b.ExpectException)
+		}
+
+		// validate RLP decoding by checking all values against test file JSON
+		if err = validateHeader(b.BlockHeader, cb.Header()); err != nil {
+			return nil, fmt.Errorf("deserialised block header validation failed: %v", err)
+		}
+		validBlocks = append(validBlocks, b)
+		blocks = append(blocks, cb)
+	}
+
+	// Insert the blocks into the chain
+	i, err := sut.InsertChain(ctx, tb, blocks)
```

Comment:
in the eth tests they were inserting a single block > execute > validate header and then continue to other blocks. Here we rather insert all collect validated blocks > enqueue > execute all of them and then return.